finetrainers/ltx_video/ltx_video_lora.py

```diff
@@ -165,7 +221,90 @@ def prepare_latents(
+    generator: Optional[torch.Generator] = None,
+    precompute: bool = False,
+) -> torch.Tensor:
+    device = device or vae.device
+
+    if image_or_video.ndim == 4:
+        image_or_video = image_or_video.unsqueeze(2)
+    assert image_or_video.ndim == 5, f"Expected 5D tensor, got {image_or_video.ndim}D tensor"
+
+    if image_or_video.size(2) == 1:
+        logger.warning("Received a single frame for LTX Video I2V training. Duplicating the frame to create a video.")
+        # Duplicate the image to create atleast one frame to run prediction on
+        image_or_video = image_or_video.repeat(1, 1, 2, 1, 1)
+
+    image_or_video = image_or_video.to(device=device, dtype=vae.dtype)
+    image_or_video = image_or_video.permute(0, 2, 1, 3, 4).contiguous()  # [B, C, F, H, W] -> [B, F, C, H, W]
+
+    # Note: we separately encode the image and video because there is a 4x compression applied. We only want to condition
+    # on the first frame of the video, and not the first 4 frames.
+    image, video = image_or_video[:, :, :1], image_or_video[:, :, 1:]
+
+    if not precompute:
+        image_latents = vae.encode(image).latent_dist.sample(generator=generator)
+        video_latents = vae.encode(video).latent_dist.sample(generator=generator)
+        latents = torch.cat([image_latents, video_latents], dim=2)
+        latents = latents.to(dtype=dtype)
+        _, _, num_frames, height, width = latents.shape
+        latents = _normalize_latents(latents, vae.latents_mean, vae.latents_std)
+        latents = _pack_latents(latents, patch_size, patch_size_t)
+        return {"latents": latents, "num_frames": num_frames, "height": height, "width": width}
+    else:
+        # Note: we separately encode the image and video because there is a 4x compression applied. We only want to condition
+        # on the first frame of the video, and not the first 4 frames.
+        image, video = image_or_video[:, :, :1], image_or_video[:, :, 1:]
+        if vae.use_slicing and image_or_video.shape[0] > 1:
+            encoded_slices_image = [vae._encode(x_slice) for x_slice in image.split(1)]
+            encoded_slices_video = [vae._encode(x_slice) for x_slice in video.split(1)]
+            h_image = torch.cat(encoded_slices_image)
+            h_video = torch.cat(encoded_slices_video)
+            h = torch.cat([h_image, h_video], dim=2)
+        else:
+            image, video = image_or_video[:, :, :1], image_or_video[:, :, 1:]
+            h_image = vae._encode(image)
+            h_video = vae._encode(video)
+            h = torch.cat([h_image, h_video], dim=2)
+
+        _, _, num_frames, height, width = h.shape
+
+        # TODO(aryan): This is very stupid that we might possibly be storing the latents_mean and latents_std in every file
+        # if precomputation is enabled. We should probably have a single file where re-usable properties like this are stored
+        # so as to reduce the disk memory requirements of the precomputed files.
```

Comment:
Anything to make it simple for now is welcome. It's a very tiny tensor so, very much okay with this